QSPIFBlockDevice.cpp

```diff
@@ -0,0 +1,1260 @@
+
+// Debug Printouts
+#define QSPIF_DEBUG_ERROR   1
+#define QSPIF_DEBUG_WARNING 2
+#define QSPIF_DEBUG_INFO    3
+#define QSPIF_DEBUG_DEBUG   4
+#define QSPIF_DEBUG_TRACE   5
+
+namespace mbed {
+
+enum qspif_default_instructions {
+    QSPIF_NOP  = 0x00, // No operation
+    QSPIF_PP = 0x02, // Page Program data
+    QSPIF_READ = 0x03, // Read data
+    QSPIF_SE   = 0x20, // 4KB Sector Erase
+    QSPIF_SFDP = 0x5a, // Read SFDP
+
+    QSPIF_WRSR = 0x01, // Write Status/Configuration Register
+    QSPIF_WRDI = 0x04, // Write Disable
+    QSPIF_RDSR = 0x05, // Read Status Register
+    QSPIF_WREN = 0x06, // Write Enable
+
+    QSPIF_RSTEN = 0x66, // Reset Enable
+    QSPIF_RST = 0x99, // Reset
+    QSPIF_RDID = 0x9f, // Read Manufacturer and JDEC Device ID
+};
+
+// Remove Remark To Enalbe Logs
+//#define QSPIF_ENABLE_LOGS 1
+
+#ifdef QSPIF_ENABLE_LOGS
+
+#define QSPIF_DEFAULT_DEBUG_LEVEL  QSPIF_DEBUG_WARNING
+
+#define __FILENAME__ (strrchr(__FILE__, '/') ? strrchr(__FILE__, '/') + 1 : __FILE__)
+
+#define QSPIF_LOG(level,...) {\
+    	if (level <= QSPIF_DEFAULT_DEBUG_LEVEL) {\
+    	  char str[256];\
+      	  sprintf(str, "\n[%s][%s:%d], ", __FILENAME__, __FUNCTION__, __LINE__);\
+            sprintf(str+strlen(str),__VA_ARGS__);\
+      	  printf(str);\
+    	}\
+    }
+#else
+#define QSPIF_LOG(level,...)
+#endif //QSPIF_ENABLE_LOGS
+
+// Mutex is used for some QSPI Driver commands that must be done sequentially with no other commands in between
+// e.g. (1)Set Write Enable, (2)Program, (3)Wait Memory Ready
+SingletonPtr<PlatformMutex> QSPIFBlockDevice::_mutex;
+
+// Local Function
+int localMathPower(int base, int exp);
+
+/********* Public API Functions *********/
+/****************************************/
+
+QSPIFBlockDevice::QSPIFBlockDevice(PinName io0, PinName io1, PinName io2, PinName io3, PinName sclk, PinName csel,
+                                   qspif_polarity_mode clock_mode, int freq)
+    : _qspi(io0, io1, io2, io3, sclk, csel, clock_mode), _device_size_bytes(0)
+{
+    _is_initialized = false;
+    _min_common_erase_size = 0;
+    _regions_count = 1;
+    _region_erase_types_bitfield[0] = ERASE_BITMASK_NONE;
+
+    //Default Bus Setup 1_1_1 with 0 dummy and mode cycles
+    _inst_width = QSPI_CFG_BUS_SINGLE;
+    _address_width = QSPI_CFG_BUS_SINGLE;
+    _address_size = QSPI_CFG_ADDR_SIZE_24;
+    _data_width = QSPI_CFG_BUS_SINGLE;
+    _dummy_and_mode_cycles = 0;
+
+    if (QSPI_STATUS_OK != _qsp_set_frequency(freq)) {
+        QSPIF_LOG(QSPIF_DEBUG_ERROR, "ERROR: QSPI Set Frequency Failed");
+    }
+}
+
+
+int QSPIFBlockDevice::init()
+{
+
+    uint8_t vendor_device_ids[4];
+    size_t data_length = 3;
+
+    _mutex->lock();
+    if (_is_initialized == true) {
+        _mutex->unlock();
+        return 0;
+    }
+
+    // Soft Reset
+    if ( -1 == _reset_flash_mem()) {
+        QSPIF_LOG(QSPIF_DEBUG_INFO, "ERROR: init - Unable to initialize flash memory, tests failed\n");
+        return QSPIF_BD_ERROR_DEVICE_ERROR;
```

Comment:
Forgot to unlock the mutex (here, and in all following error cases).

Reply:
missed this one - Thanks! I'll add an exit point for errors with unlocking